core/templates/services/math-interactions.service.ts

```diff
@@ -524,6 +524,42 @@ export class MathInteractionsService {
   // The input terms to this function should be the terms split by '+'/'-'
   // from an expression.
   termsMatch(term1: string, term2: string): boolean {
+    let getConstants = function(expression: string): string[] {
+      let constants = expression.match(/(\d*\.\d+)|(\d+)/g);
+      if (constants !== null) {
+        return constants;
+      }
+      return [];
+    };
+
+    let partsMatch = function(part1: string, part2: string): boolean {
+      // Nerdamer simplifies constant terms automatically so to avoid that,
+      // we'll replace all constants with a variable.
+      let usedVariables = nerdamer(part1).variables().concat(
+        nerdamer(part2).variables());
+      let usedVariablesSet = new Set(usedVariables);
+
+      let usedConstants = getConstants(part1).concat(getConstants(part2));
+      // Sorting in reverse by length so that longer strings get replaced before
+      // shorter ones so that overlapping substrings don't cause conflicts.
+      usedConstants.sort((a, b) => b.length - a.length);
+
+      let charCode = 'a'.charCodeAt(0);
+      for (let i = 0; i < usedConstants.length; i++) {
+        let constant = usedConstants[i];
+        if (i === 0 || constant !== usedConstants[i - 1]) {
+          let letter = String.fromCharCode(charCode);
+          while (usedVariablesSet.has(letter)) {
+            letter = String.fromCharCode(++charCode);
```

Comment:
Can this be broken by an input string that has all of the alphabetic chars? (Maybe add a test case for this.)

Reply:
Yes, it will. But the new approach won't fail in this case. Test case added.